rhine-terminal/TerminalSimple.hs

```diff
@@ -0,0 +1,113 @@
+{- | Example application for the @rhine-terminal@ library. -}
+
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE NamedFieldPuns #-}
+
+-- base
+import Prelude hiding (putChar)
+import System.Exit (exitSuccess)
+import System.IO hiding (putChar)
+import Control.Monad
+
+-- text
+import Data.Text (Text)
+import qualified Data.Text as T
+-- terminal
+import System.Terminal
+import System.Terminal.Internal
+-- rhine
+import FRP.Rhine
+-- rhine-terminal
+import FRP.Rhine.Terminal
+
+type App = TerminalT LocalTerminal IO
+
+-- Clocks
+
+data Input
+  = Char Char Modifiers
+  | Space
+  | Backspace
+  | Enter
+  | Exit
+
+type InputClock = SelectClock TerminalEventClock Input
+
+inputClock :: InputClock
+inputClock = SelectClock
+  { mainClock = TerminalEventClock
+  , select }
+    where
+      select :: Tag TerminalEventClock -> Maybe Input
+      select = \case
+        Right (KeyEvent (CharKey k) m)  -> Just (Char k m)
+        Right (KeyEvent SpaceKey _)     -> Just Space
+        Right (KeyEvent BackspaceKey _) -> Just Backspace
+        Right (KeyEvent EnterKey _)     -> Just Enter
+        Left _                          -> Just Exit
+        _                               -> Nothing
+
+type PromptClock = LiftClock IO (TerminalT LocalTerminal) (Millisecond 1000)
+
+type AppClock = ParallelClock App InputClock PromptClock
+
+-- ClSFs
+
+inputSource :: ClSF App InputClock () Input
+inputSource = tagS
+
+promptSource :: ClSF App PromptClock () Text
+promptSource = flip T.cons " > " . (cycle " ." !!) <$> count
+
+inputSink ::  ClSF App cl Input ()
+inputSink = arrMCl $ \case
+  -- Don't display Ctrl-J https://github.com/lpeterse/haskell-terminal/issues/17
+  Char c m  -> when (c /= 'J' && m /= ctrlKey) $ putChar c >> flush
```

Comment:
```suggestion
  Char 'J' Ctrl  -> flush
  Char c m  -> putChar c >> flush
```
I think this is the clearest and most obviously correct solution.

Reply:
Unfortunately, we [can't use patterns](https://github.com/lpeterse/haskell-terminal/issues/14) and `Modifiers` are not exported, so we can't just pattern match :/ I'll probably add this to my fork at some point...